pt/lessons/advanced/gen-stage.md

```diff
@@ -0,0 +1,260 @@
+---
+version: 1.0.1
+title: GenStage
+redirect_from:
+  - /lessons/advanced/gen-stage/
+---
+
+Nesta lição, vamos examinar de perto o GenStage, para que serve e como podemos usá-lo em nossas aplicações.
+
+{% include toc.html %}
+
+## Introdução
+
+Então, o que é GenStage? De acordo com a documentação oficial, ele é uma "especificação e um fluxo computacional para o Elixir", mas o que isso significa pra gente?
+
+Significa que o GenStage nos fornece uma forma de definir um pipeline de trabalho a ser realizado por passos independentes (ou etapas) em processos separados; se você já trabalhou com pipelines anteriormente, então alguns desses conceitos devem ser familiares.
+
+Para entender melhor como isso funciona, vamos visualizar um simples fluxo produtor-consumidor:
+
+```
+[A] -> [B] -> [C]
+```
+
+Neste exemplo temos três etapas: `A` um produtor, `B` um produtor-consumidor, e `C` um consumidor.  `A` produz um valor que é consumido por `B`, `B` executa algum trabalho e retorna um novo valor que é recebido pelo nosso consumidor `C`; o papel da nossa etapa é importante como veremos na próxima seção.
+
+Enquanto nosso exemplo é 1-para-1, produtor-para-consumidor, é possível que ambos tenham múltiplos produtores e múltiplos consumidores em qualquer etapa.
+
+Para ilustrar melhor esses conceitos, vamos construtir um pipeline com GenStage, mas antes vamos explorar os papéis que o GenStage depende um pouco mais.
+
+## Consumidores e Produtores
+
+Conforme lemos, o papel que damos à nossa etapa é importante. A especificação do GenStage reconhece três papéis:
+
++ `:producer` — Uma fonte. Produtores esperam por demanda de consumidores e respondem com os eventos solicitados.
+
++ `:producer_consumer` — Tanto uma fonte como um destino. Produtor-consumidores podem responder por demandas de outros consumidores assim como solicitar eventos de produtores.
+
++ `:consumer` — A sink. Um consumidor solicita e receve dados de produtores.
```

Comment:
I believe we could translate "sink" to "tanque" or "reservatório". There is also a typo in "receve" word; it should read "recebe" instead.